### PR TITLE
Aggregate in Chunks

### PR DESCRIPTION
**Pull request by steremma**
This work aggregates the new huge dataset in reasonable time using chunk processing. We only use a few columns which are split into dynamic (repeated for each month) and static (only once per user) features.

The main operations are grouping and pivoting. The whole thing is expected to take around 50 minutes on a commodity machine the first time it runs. Each experiment thereafter should just reload the produced CSV files.

### CODE REVIEW

File: google_analytics/aggregation.py [hunk 1]
```diff
@@ -1,7 +1,138 @@
...
 import warnings
-import numpy as np
+
 import pandas as pd
+import numpy as np
+
+
+def reduce_df(path, output, nrows=None, chunksize=20000):
+    """ Load Google analytics data from JSON into a Pandas.DataFrame. """
+    if nrows and chunksize:
```
Comment (Christiannewisse) [inline google_analytics/aggregation.py]: Small remark: this message is now not printed when nrows = None (i.e the whole dataset).

File: google_analytics/aggregation.py [hunk 2]
```diff
@@ -1,7 +1,138 @@
...
+        msg = "Reading {} rows in chunks of {}. We are gonna need {} chunks"
+        print(msg.format(nrows, chunksize, nrows / chunksize))
+
+    temp_dir = "../data/temp"
```
Comment (Christiannewisse) [inline google_analytics/aggregation.py]: We are not 'cleaning' this directory. So right now we are loading the train set in 86 chucks and saving them in the folder temp. Then we load everything in this folder and merge it. Here everything goes fine. Then we are loading the test set in 21 chucks and saving them in the same temp folder. Basically, we are writing over the first 21 chucks of the train set. Then we are concatenating everything in the folder temp, which includes both the test set (chunk 0 till 20) and the last part of the train set (chunck 21 till 86). 
So basically, we need to empty the temp folder before starting with a new dataset.

File: google_analytics/aggregation.py [hunk 3]
```diff
@@ -1,7 +1,138 @@
...
+    if not os.path.exists(temp_dir):
+        os.makedirs(temp_dir)
+
+    JSON_COLUMNS = ['device', 'geoNetwork', 'totals', 'trafficSource']
+
+    i = 0
+    for chunk in pd.read_csv(path,
```
Comment (Christiannewisse) [inline google_analytics/aggregation.py]: Nice work with the chunks!

File: google_analytics/aggregation.py [hunk 4]
```diff
@@ -1,7 +1,138 @@
...
+    df['transactions'].fillna(0, inplace=True)
+    df['transactionRevenue'].fillna(0.0, inplace=True)
+
+    def agg_date(s):
+        date = pd.datetime.strptime(s, '%Y-%m-%d')
+        return "{}_{}".format(date.month, date.year)
+
+    df["date"] = df["date"].apply(agg_date)
+
+    def most_frequent(x):
```
Comment (Christiannewisse) [inline google_analytics/aggregation.py]: What is the benefit of only including the most frequent value? In the end we still need to apply OHE right, so we still end up with a column for each country.

File: google_analytics/aggregation.py [hunk 5]
```diff
@@ -1,7 +1,138 @@
...
+    print("First grouping, this will take a while")
+    grouped = df.groupby(["fullVisitorId", "date"], as_index=False).agg(agg)
+
+    # Long to wide pivot - one row per visitor.
+    dynamic_columns = ['transactionRevenue', 'visits', 'transactions', 'pageviews']
+    wide = pd.pivot_table(grouped, index='fullVisitorId', columns='date', values=dynamic_columns)
+
+    # Collapse multi-index.
+    wide.columns = wide.columns.to_series().str.join('_')
+    wide.reset_index(inplace=True)
```
Comment (Christiannewisse) [inline google_analytics/aggregation.py]: I am not sure but shouldn't the NaN values be filled with zeros?

File: google_analytics/notebooks/dynamic_aggregation.ipynb [hunk 6]
```diff
@@ -33,9 +43,69 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "deletable": true,
+    "editable": true
+   },
+   "outputs": [],
+   "source": [
+    "print(\"Let's widen the train dataset.\")\n",
```
Comment (Christiannewisse) [inline google_analytics/notebooks/dynamic_aggregation.ipynb]: Now we aggregate based on the given train and test periods. However, if we want to change the periods of train_x, train_y, test_x, and test_y, maybe it is good to use the function 'create_train_test' before doing the aggregation. This function is splitting the train and test as given by Kaggle into train and test sets based on a given period.  Furthermore, in this function visitors that visited once and never bought during the training period are dropped. If I am right, this is not done yet. 
To summarize my thoughts a bit, I was thinking about the following steps
1. Load train_v2 and test_v2.
2. Drop unnecessary columns.
3. Respecify train_x, train_y, test_x, and test_y based on given periods. 
4. Drop visitors that visit once and did not buy in x from both x and y.
5. Aggregate the four datasets per visitor (for x: OHE or boolean (for example isUSA) for static and create dynamic features; for y: sum of the revenue)
6. Keep only columns in train_x and test_x that appear in both datasets. 
I am not sure if my thoughts are right, but are this the right steps, do I miss something, or are there maybe unnecessary steps?

File: google_analytics/aggregation.py (hunk 2 above)
Comment (Christiannewisse) [reply google_analytics/aggregation.py]: This should do the job:     
for the_file in os.listdir(temp_dir):
        try:
            os.unlink(os.path.join(temp_dir, the_file))
        except Exception as e:
            print(e)

File: google_analytics/aggregation.py [hunk 7]
```diff
@@ -1,7 +1,138 @@
...
+        # Normalize JSON columns
+        for column in JSON_COLUMNS:
+            column_as_df = pd.io.json.json_normalize(chunk[column])
+            chunk = chunk.drop(column, axis=1).merge(column_as_df, right_index=True, left_index=True)
+
+        # Parse date
+        chunk['date'] = chunk['date'].apply(lambda x: pd.datetime.strptime(str(x), '%Y%m%d'))
+
+        # Only keep relevant columns
+        cols = ['date', 'fullVisitorId', 'operatingSystem', 'country', 'browser',
```
[the remark this answers is not in the document]
Comment (steremma) [reply google_analytics/aggregation.py]: yeap we should

File: google_analytics/aggregation.py (hunk 1 above)
Comment (steremma) [reply google_analytics/aggregation.py]: True but I was too bored to fix. Will do!

File: google_analytics/aggregation.py (hunk 2 above)
Comment (steremma) [reply google_analytics/aggregation.py]: thats a huge bug that I didnt notice thanks!

File: google_analytics/aggregation.py (hunk 4 above)
Comment (steremma) [reply google_analytics/aggregation.py]: actually...you are right :P For the countyr it does because I was thinking of replacing it with the financial metrics but for the rest maybe I should OHE directly.

File: google_analytics/aggregation.py (hunk 5 above)
Comment (steremma) [reply google_analytics/aggregation.py]: Sure, we can do it here or later (after the `wide` dataset is loaded)

File: google_analytics/notebooks/dynamic_aggregation.ipynb (hunk 6 above)
Comment (steremma) [reply google_analytics/notebooks/dynamic_aggregation.ipynb]: I was thinking that setting a time frame is now as simple as dropping certain columns (corresponding to the dates outside the time-frame you want). But I might be missing something, let's discuss this in person

File: google_analytics/aggregation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (steremma) [reply google_analytics/aggregation.py]: but then how would we aggregate? I mean a visitor can have multiple `visitStartTime` values, one for each section. Perhaps aggregate by concatenation so that the grouped field is a list of all `visitStartTime`s?
